zk/stages/stage_sequence_execute_utils.go

```diff
@@ -0,0 +1,324 @@
+	engine        consensus.Engine
+	zkVmConfig    *vm.ZkConfig
+	badBlockHalt  bool
+	stateStream   bool
+	accumulator   *shards.Accumulator
+	blockReader   services.FullBlockReader
+
+	dirs      datadir.Dirs
+	historyV3 bool
+	syncCfg   ethconfig.Sync
+	genesis   *types.Genesis
+	agg       *libstate.AggregatorV3
+	zk        *ethconfig.Zk
+
+	txPool   *txpool.TxPool
+	txPoolDb kv.RwDB
+}
+
+func StageSequenceBlocksCfg(
+	db kv.RwDB,
+	pm prune.Mode,
+	batchSize datasize.ByteSize,
+	changeSetHook ChangeSetHook,
+	chainConfig *chain.Config,
+	engine consensus.Engine,
+	vmConfig *vm.ZkConfig,
+	accumulator *shards.Accumulator,
+	stateStream bool,
+	badBlockHalt bool,
+
+	historyV3 bool,
+	dirs datadir.Dirs,
+	blockReader services.FullBlockReader,
+	genesis *types.Genesis,
+	syncCfg ethconfig.Sync,
+	agg *libstate.AggregatorV3,
+	zk *ethconfig.Zk,
+
+	txPool *txpool.TxPool,
+	txPoolDb kv.RwDB,
+) SequenceBlockCfg {
+	return SequenceBlockCfg{
+		db:            db,
+		prune:         pm,
+		batchSize:     batchSize,
+		changeSetHook: changeSetHook,
+		chainConfig:   chainConfig,
+		engine:        engine,
+		zkVmConfig:    vmConfig,
+		dirs:          dirs,
+		accumulator:   accumulator,
+		stateStream:   stateStream,
+		badBlockHalt:  badBlockHalt,
+		blockReader:   blockReader,
+		genesis:       genesis,
+		historyV3:     historyV3,
+		syncCfg:       syncCfg,
+		agg:           agg,
+		zk:            zk,
+		txPool:        txPool,
+		txPoolDb:      txPoolDb,
+	}
+}
+
+type stageDb struct {
+	tx          kv.RwTx
+	hermezDb    *hermez_db.HermezDb
+	eridb       *db2.EriDb
+	stateReader *state.PlainStateReader
+	smt         *smtNs.SMT
+}
+
+func newStageDb(tx kv.RwTx) *stageDb {
+	sdb := &stageDb{
+		tx:          tx,
+		hermezDb:    hermez_db.NewHermezDb(tx),
+		eridb:       db2.NewEriDb(tx),
+		stateReader: state.NewPlainStateReader(tx),
+		smt:         nil,
+	}
+	sdb.smt = smtNs.NewSMT(sdb.eridb)
+
+	return sdb
+}
+
+func prepareForkId(cfg SequenceBlockCfg, lastBatch, executionAt uint64, hermezDb *hermez_db.HermezDb) (uint64, error) {
+	var forkId uint64 = 0
+	var err error
+
+	if executionAt == 0 {
+		// capture the initial sequencer fork id for the first batch
+		forkId = cfg.zk.SequencerInitialForkId
+		if err := hermezDb.WriteForkId(1, forkId); err != nil {
+			return forkId, err
+		}
+		if err := hermezDb.WriteForkIdBlockOnce(uint64(forkId), 1); err != nil {
+			return forkId, err
+		}
+	} else {
+		forkId, err = hermezDb.GetForkId(lastBatch)
+		if err != nil {
+			return forkId, err
+		}
+		if forkId == 0 {
+			return forkId, errors.New("the network cannot have a 0 fork id")
+		}
+	}
+
+	return forkId, nil
+}
+
+func prepareHeader(tx kv.RwTx, previousBlockNumber, deltaTimestamp, forkId uint64, coinbase common.Address) (*types.Header, *types.Block, error) {
+	parentBlock, err := rawdb.ReadBlockByNumber(tx, previousBlockNumber)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	useTimestampOffsetFromParentBlock := deltaTimestamp != math.MaxUint64
```

Comment:
a comment here could be helpful, I had to double check where the logic would cause a max uint64 to be used.